### PR TITLE
Properly detect hex-colors in html5-css hex-color waypoints

### PR DESCRIPTION
**Pull request by DDaems**
Origenally you could pass all tests without ever placing the
hex code in the correct place, als abbreviates didn't work either
for background and hex code
- Can use the shortcode to pass the test now.
- Atleasts needs a body present with curly brackets
- Needs a background property being present in the body element
- The background property must be followed by a the proper hex color

Fixes a problem that was reported in the following issue
#5010

### CODE REVIEW

**Comment by SaintPeter** [comment]
I am unable to seed, I'm getting an error:
`error [SyntaxError: c:\mean\FreeCodeCamp\seed\challenges\html5-and-css.json: Unexpected token {]`

I suspect there is an escaping problem with the new regex.  You need to double-escape them.

You may try installing https://github.com/FreeCodeCamp/COM1000 which does the escaping for you.  I was unable to load this file even with it, though, so you'll have to start from scratch.

**Comment by bugron** [comment]
Even GitHub highlights the corrupted part. Also change the title to something more readable and add `closes #5010` to the description, thanks.

**Comment by DDaems** [comment]
Hold on on merging this, i changed the first test to be a bit more explicit. I'm trying to alter the second one a bit more. To make sure the hex is behind a color background tag within the body

**Comment by DDaems** [comment]
@SaintPeter Thanks bookmarked, might check it out later.
@bugron Done here, the end result is really rewarding and i learned some intresting stuff.
``` 
// PASSES
h1{ color: blue; } 
body{ color: red; background: #ffF; margin: 15px; font-size: 15px }
```
```
// FAILS
body{ color: red; background: #ffFf; margin: 15px; font-size: 15px }
```

File: seed/challenges/html5-and-css.json [hunk 1]
```diff
@@ -3775,8 +3775,8 @@
         "Replace the word <code>black</code> in our <code>body</code> element's background-color with its <code>hex code</code> representation, <code>#000000</code> "
       ],
       "tests": [
-        "assert($(\"body\").css(\"background-color\") === \"rgb(0, 0, 0)\", 'message: Give your <code>body</code> element the background-color of black.');",
-        "assert(code.match(/#000000/g) && code.match(/#000000/g).length > 0, 'message: Use the <code>hex code</code> for the color black instead of the word <code>black</code>. For example <code>body { color&#58; #000000; }</code>');"
+        "assert($(\"body\").css(\"background-color\") === \"rgb(0, 0, 0)\" && code.match(/body\\s?{([\\s\\S]*?)}/i).toString().indexOf(\"background\") != -1, 'message: Give your <code>body</code> element the background-color of black.');",
```
Comment (bugron) [inline seed/challenges/html5-and-css.json]: Why are you matching here the word `background` if you're doing the some thing in the next test? 
`$(\"body\").css(\"background-color\") === \"rgb(0, 0, 0)\"` does exactly what the test message is talking about.

File: seed/challenges/html5-and-css.json [hunk 2]
```diff
@@ -3775,8 +3775,8 @@
...
+        "assert(code.match(/body\\s?{([\\s\\S]*?)}/i)[0].match(/background.*/i)[0].match(/(^#|#)([a-zA-Z0-9]+)/gi).toString().toUpperCase() === \"#000000\" || code.match(/body\\s?{([\\s\\S]*?)}/i)[0].match(/background.*/i)[0].match(/(^#|#)([a-zA-Z0-9]+)/gi).toString().toUpperCase() === \"#000\", 'message: Use the <code>hex code</code> for the color black instead of the word <code>black</code>. For example <code>body { color&#58; #000000; }</code>');"
```
Comment (bugron) [inline seed/challenges/html5-and-css.json]: Three `match`es and two separate tests for `#000000` and `#000`? For example, the following Regexp does the same thing: `"body {background-color: #000;}".match(/body\s*{\s*background.*\s*:\s*#000(000)?;\s*}/i)`. This won't be `null` only if your `body` element's `background-color` is black (`#000` or `#000000`). There is no need even explicitly check for equality here because `Chai.js` coerces the result to `Boolean` for you.
I've verified in my local that my regex works properly so, please, `amend` your commit and apply these changes to the tests, thanks.

File: seed/challenges/html5-and-css.json (hunk 1 above)
Comment (DDaems) [inline seed/challenges/html5-and-css.json]: I had written te test case origenally to fix `Use Hex Code to Color Elements White`.
The default background-color of the page is `#fff` wich would pass even if there was no code written. Thats the reason why i made the test a bit more explicit.

I simple copy pasted my code afterwards and forgot that the others do not need that test.

File: seed/challenges/html5-and-css.json (hunk 2 above)
Comment (DDaems) [inline seed/challenges/html5-and-css.json]: > `match(/body\s*{[\\s\\S]*?background.*\s*:\s*#000(000)?;[\\s\\S]*?}/gi)`

Ah thanks, I knew it was possible. I couldn't translate the words in my mind to a term for the searchengine. I ended up doing seperate checks :(.

**Comment by DDaems** [comment]
@bugron I"m done here, I updated all test. And altered your code to be more robust. Because it was failing some of my testcases. 

http://codepen.io/DDaems/pen/JGoRop

File: seed/challenges/html5-and-css.json [hunk 3]
```diff
@@ -3814,8 +3814,8 @@
         "Let's turn our <code>body</code> element's background-color white by changing its hex code to <code>#FFFFFF</code>"
       ],
       "tests": [
-        "assert($(\"body\").css(\"background-color\") === \"rgb(255, 255, 255)\", 'message: Your <code>body</code> element should have the <code>background-color</code> of white.');",
-        "assert(code.match(/#FFFFFF/ig) && code.match(/#FFFFFF/ig).length > 0, 'message: Use the <code>hex code</code> for the color white instead of the word <code>white</code>. For example <code>body { color&#58; #FFFFFF; }</code>');"
+        "assert($(\"body\").css(\"background-color\") === \"rgb(255, 255, 255)\" && code.match(/body\\s*{[\\s\\S]*?background.*[\\s\\S]*?}/gi), 'message: Your <code>body</code> element should have the <code>background-color</code> of white.');",
```
Comment (bugron) [inline seed/challenges/html5-and-css.json]: I think you forgot here to remove `&& code.match(/body\\s*{[\\s\\S]*?background.*[\\s\\S]*?}/gi)`.

File: seed/challenges/html5-and-css.json [hunk 4]
```diff
@@ -3776,7 +3776,7 @@
       ],
       "tests": [
         "assert($(\"body\").css(\"background-color\") === \"rgb(0, 0, 0)\", 'message: Give your <code>body</code> element the background-color of black.');",
-        "assert(code.match(/#000000/g) && code.match(/#000000/g).length > 0, 'message: Use the <code>hex code</code> for the color black instead of the word <code>black</code>. For example <code>body { color&#58; #000000; }</code>');"
+        "assert(code.match(/body\\s*{(([\\s\\S]*;\\s*?)|\\s*?)background.*\\s*:\\s*?#000(000)?((\\s*})|(;[\\s\\S]*?}))/gi), 'message: Use the <code>hex code</code> for the color black instead of the word <code>black</code>. For example <code>body { color&#58; #000000; }</code>');"
```
Comment (bugron) [inline seed/challenges/html5-and-css.json]: Why is this here: `(([\\s\\S]*;\\s*?)|\\s*?)` and `((\\s*})|(;[\\s\\S]*?}))`? Using too many capture groups is not a good idea.

**Comment by bugron** [comment]
@FreeCodeCamp/issue-moderators can you, please, share your thoughts about this PR. I'll left this PR for you to QA.

File: seed/challenges/html5-and-css.json (hunk 3 above)
Comment (DDaems) [inline seed/challenges/html5-and-css.json]: I intentionally left that in so the first test wouldn't pass automaticly, even if there was no code written. Because the default background is allready white.

Shall i remove this?

File: seed/challenges/html5-and-css.json (hunk 4 above)
Comment (DDaems) [inline seed/challenges/html5-and-css.json]: to pass:
```
body { background: #fff }
body { color: red; background: #fff }
```
2fail:
```
body { background: #ffffffaefze }
body { color: red background: #fff }
```

File: seed/challenges/html5-and-css.json (hunk 3 above)
Comment (ltegman) [inline seed/challenges/html5-and-css.json]: The seed code has a starting background-color of `#000000`. This should be removed.

Comment (DDaems) [inline seed/challenges/html5-and-css.json]: Apperently i was wrong, the default is black as in the seed yes. I allways forget to reset the challenge ;-) a reload is not enough. I'll remove it. the seed can stay imo

**Comment by ltegman** [comment]
There are now merge conflicts that will need to be resolved before this can be merged :/

**Comment by DDaems** [comment]
@ltegman resolved conflicts

**Comment by ltegman** [comment]
Not seeing any issues with this and given that no one else has left any feedback I'm going to merge. Thanks! :+1:
